File: java_solutions/747.largest-number-at-least-twice-of-others.java
/*
 * @lc app=leetcode id=747 lang=java
 *
 * [747] Largest Number At Least Twice of Others
 */

class Solution {
    public int dominantIndex(int[] nums) {
        return IntStream.range(0, nums.length) .boxed() .reduce((a, b) -> nums[a] > nums[b] ? a : b) .map(maxIdx -> Arrays.stream(nums).allMatch(n -> n == nums[maxIdx] || nums[maxIdx] >= 2 * n) ? maxIdx : -1) .orElse(-1);
    }
}

File: kotlin_solutions/747.largest-number-at-least-twice-of-others.kt
/*
 * @lc app=leetcode id=747 lang=kotlin
 *
 * [747] Largest Number At Least Twice of Others
 */
class Solution { fun dominantIndex(nums: IntArray): Int = nums.indices.maxByOrNull { nums[it] }!!.let { maxIdx -> if (nums.all { it == nums[maxIdx] || nums[maxIdx] >= 2 * it }) maxIdx else -1 } }
